flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/FlinkKafkaProducerITCase.java

```diff
@@ -597,6 +602,93 @@ public void testMigrateFromAtExactlyOnceToAtLeastOnce() throws Exception {
         deleteTestTopic(topic);
     }
 
+    @Test
+    public void testCustomizeTransactionalIdPrefix() throws Exception {
+        String transactionalIdPrefix = "my-prefix";
+
+        Properties properties = createProperties();
+        String topic = "testCustomizeTransactionalIdPrefix";
+        FlinkKafkaProducer<Integer> kafkaProducer =
+                spy(
+                        new FlinkKafkaProducer<>(
+                                topic,
+                                integerKeyedSerializationSchema,
+                                properties,
+                                FlinkKafkaProducer.Semantic.EXACTLY_ONCE));
+        kafkaProducer.setTransactionalIdPrefix(transactionalIdPrefix);
+
+        try (OneInputStreamOperatorTestHarness<Integer, Object> testHarness =
+                new OneInputStreamOperatorTestHarness<>(
+                        new StreamSink<>(kafkaProducer), IntSerializer.INSTANCE)) {
+            testHarness.setup();
+            testHarness.open();
+            testHarness.processElement(1, 0);
+            testHarness.snapshot(0, 1);
+        }
+
+        ArgumentCaptor<FlinkKafkaProducer.KafkaTransactionState> stateCaptor =
+                ArgumentCaptor.forClass(FlinkKafkaProducer.KafkaTransactionState.class);
+        verify(kafkaProducer).invoke(stateCaptor.capture(), any(), any());
```

Comment:
We are [discouraging use of mockito](https://flink.apache.org/contributing/code-style-and-quality-common.html#design-for-testability). I see at least a couple ways how to avoid it:

1. extend `FlinkKafkaProducer` and override `beginTransaction()` that would proxy the call to the super class and capture the result
2. use `org.apache.flink.streaming.api.functions.sink.TwoPhaseCommitSinkFunction#currentTransaction()` existing method to validate it (it's a `protected` method so you would need to access it in test from the same package?

Reply:
I thought it would be hard to verify the actually-used transactional.id except using mockito. The ways you shared are really helpful! Let me fix it.

Reply:
👍 
3. you could also provide package private (it shouldn't be exposed to the users) method similar to `org.apache.flink.streaming.connectors.kafka.FlinkKafkaProducer#getTransactionCoordinatorId`